tools/mason/MasonArgParse.chpl

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+module MasonArgParse {
+  private use List;
+  private use Map;
+  private use IO;
+  private use Sort;
+
+  const DEBUG=false;
+  // TODO: Implement required/optional flag
+  // TODO: Implement default values for optional opts
+  // TODO: Verify no duplicate names, flags defined by dev
+  // TODO: Make sure we don't shadow Chapel flags
+  // TODO: Make sure we don't shadow config vars  
+  // TODO: Implement Help message and formatting
+  // TODO: Add bool flags
+  // TODO: Add int opts
+  // TODO: Add positional arguments
+  // TODO: Add public github issue when available
+
+  if chpl_warnUnstable then
+    compilerWarning("ArgumentParser is unstable.");
+
+// A generic argument parser error
+  class ArgumentError : Error {
+    var msg:string;
+    proc init(msg:string) {
+      this.msg = msg;
+    }
+    override proc message() {
+      return msg;
+    }
+  }
+  
+// indicates a result of argument parsing
+  class Argument {
+    var present: bool=false;
+    var values: list(string);     
+    
+    proc getValue(){     
+      return this.values.first();
+    }
+    iter getValues(){
+      for val in values {
+        yield val;
+      }      
+    }    
+    proc hasValue(){
+      return !this.values.isEmpty();
+    }
+  }
+
+  // stores the definition of an option
+  class Action {
+    var name:string;
+    var numOpts:int;
+    var opts:[0..numOpts-1] string;
+    var numArgs:range;
+
+    // TODO: Decouple the argument from the action
+    // maybe pass a list to fill by reference and have the argparser populate
+    // the argument instead?
+    // also need a bool by ref to indicate presence of arg or not
+    proc match(args:[?argsD]string, startPos:int, ref myArg:Argument) throws {
```

Comment:
Since you are just mutating the `myArg:Argument` fields and it is a class, no need to write `ref myArg` here. You would need it if you made it a `record`.